### PR TITLE
Refactor repository to enable more unitary unit tests

### PR DESCRIPTION
**Pull request by nrc**
### What problem does this PR solve? 

The repository and mirror abstractions mixed IO and business logic. That means that unit tests were more like integration tests since they had to refer to files on disk and tested the whole pipeline of the repository.

### What is changed and how it works?

Refactored repository and mirror (details in first commit message). Refactored their tests to be more unit-like.

### Check List 

 - Unit test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change

### CODE REVIEW

File: pkg/repository/fileSource.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
+// Copyright 2020 PingCAP, Inc.
```
Comment (lonng) [inline pkg/repository/fileSource.go:1]: Please rename the file name to `file_source.go`, it's more Golang idiomatic.

File: pkg/repository/fileSource.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package repository
+
+import (
+	"crypto/sha1"
+	"encoding/hex"
+	"encoding/json"
+	"github.com/pingcap-incubator/tiup/pkg/utils"
```
Comment (lonng) [inline pkg/repository/fileSource.go:20]: Please separate the 3rd dependencies into the individual import group.

File: pkg/repository/fileSource_test.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
+// Copyright 2020 PingCAP, Inc.
```
Comment (lonng) [inline pkg/repository/fileSource_test.go:1]: ditto

File: pkg/repository/fileSource.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+		return errors.Trace(err)
+	}
+
+	if !expand {
+		return nil
+	}
+
+	tarFile, ok := tarReader.(*os.File)
+	if !ok {
+		return errors.Errorf("Expected file, found %v", tarReader)
```
Comment (lonng) [inline pkg/repository/fileSource.go:109]: Error string should start with lower case.

**Comment by nrc** [comment]
All comments addressed, PTAL @lonng

**Pull request merged**